src/shell_history.rs

```diff
@@ -49,80 +52,161 @@ pub fn history_file_path() -> PathBuf {
     })
 }
 
-pub fn full_history(path: &PathBuf) -> Vec<String> {
-    let history_contents = read_ignoring_utf_errors(&path);
+/// Represents each entry in a history file.
+#[derive(Debug)]
+pub struct HistoryCommand {
+    /// The user's command.
+    pub command: String,
+    /// When the command was run, in seconds since Unix epoch.
+    pub when: i64,
+    /// The format of the file, so we can write the record back out.
+    pub format: HistoryFormat,
```

Comment:
Do we need to store this on every HistoryCommand?

Reply:
We don't *have* to, but it allows us to write the command back out (to a history file, debug print, etc.) in its native format without having to pass a `HistoryFormat` around in more places, some of which didn't make sense to me.  We can't make the `Display` implementation take extra arguments, since we don't own the trait, for example.  I'm sure I could find a way to pass it around where needed if you like.

Reply:
It's likely fine. I was wondering if it's use a lot more RAM, but it's probably fine.

Reply:
As it stands now, it's a small enum containing only a bool in the fields, and I confirmed it's only 1 byte.  I agree that it's a potential concern if we add more data inside the enum later, since it's stored for each historical command, but as-is it'd be dwarfed by the size of the commands themselves and their timestamps.  If history size overall becomes a problem, it seems possible to move to a more stream-oriented model, but it'd be a bit of an overhaul.  And if it's just a problem with extra fields in `HistoryFormat` added later, we'd have the alternative of moving `HistoryFormat` out of `HistoryCommand` at that point, too, at the cost of the bit of complexity I mentioned.